### PR TITLE
Log the skipped optimizer step on gradient overflow

### LINKED ISSUE #8415: ZeRO skips the optimizer step on gradient overflow without logging anything
**Repo:** deepspeedai/DeepSpeed
**Version:** DeepSpeed 0.19.2

## Problem

When ZeRO-1/2 detects a non-finite gradient it clears gradients and returns early from `step()`
(`deepspeed/runtime/zero/stage_1_and_2.py:2154`):

```python
self._update_scale(self.overflow)
if self.overflow:
    see_memory_usage('After overflow before clearing gradients')
    self.zero_grad(set_to_none=True)
    ...
    return
```

Nothing is logged. Under fp16 this is usually survivable because `_update_scale` reports the loss
scale changing, which gives an operator an indirect signal. Under **bf16** the loss scale is fixed,
so there is no output at all: `engine.step()` returns normally, the training loop advances, the LR
scheduler steps, metrics are emitted, and the run looks completely healthy while **no weights are
ever updated**.

`DynamicLossScaler` does log `OVERFLOW! ... Skipping step`, but `CreateLossScaler` only returns it
for `fp16 + dynamic_scaling`; bf16 always gets the static `LossScaler`, which logs nothing. The
engine does count the skip in `engine.skipped_steps`, but that is only surfaced by
`_report_progress`, which is gated on `steps_per_print` — and `STEPS_PER_PRINT_DEFAULT` is `None`,
so on a default config it never fires.

In a bf16 RL run this meant every optimizer step was skipped with no indication. The only
observable symptom was that weights synced downstream were byte-identical across steps; loss and
reward curves stayed plausible throughout, since reward noise readily hides a frozen policy.

## Related

`check_grad_overflow` defaults to `False` for bf16 while fp16 forces it `True`
(`deepspeed/runtime/precision_config.py`). That default is defensible — bf16's range makes
overflow rare — but the combination is a sharp edge: with the check off, an Inf gradient flows
into the optimizer and writes NaN into every trainable tensor; with it on, steps are skipped
silently. Both failure modes are invisible. Documenting the flag's interaction with bf16 would
help even if the default stays.

### PR DESCRIPTION
**Pull request by sfc-gh-abkulkarni**
ZeRO returns early from `step()` on non-finite gradients. `DynamicLossScaler` reports that from inside `update_scale`, but `CreateLossScaler` only returns it for `fp16 + dynamic_scaling` — bf16 and fp32 always get the static `LossScaler`, whose `update_scale` is an inherited no-op.

The skip is counted in `engine.skipped_steps`, but that is only surfaced by `_report_progress`, gated on `steps_per_print`, and `STEPS_PER_PRINT_DEFAULT` is `None`. So on a default bf16 config the loop advances, the LR scheduler steps and metrics are emitted while no weights are updated, with no output at all.

Fixes #8415

### CODE REVIEW

File: deepspeed/runtime/zero/stage3.py [hunk 1]
```diff
@@ -2521,6 +2521,11 @@ def _unflatten_partitioned_parameters(self, sub_group_id):
             partitioned_param.data = q.data
 
     def _overflow_clean_up(self, prev_scale):
+        # DynamicLossScaler reports the skip itself; a static scale (always the case for bf16)
```
[the remark this answers is not in the document]
Comment (sfc-gh-abkulkarni) [reply deepspeed/runtime/zero/stage3.py]: The commit does carry the trailer — `git log` on the reviewed commit `a99600f0b9` shows `Signed-off-by: Abhishek Kulkarni <...>`, and the DCO check passes. Reading this as a false positive.

File: deepspeed/runtime/zero/stage3.py [hunk 2]
```diff
@@ -2521,6 +2521,11 @@ def _unflatten_partitioned_parameters(self, sub_group_id):
...
+        # does not, leaving the step silently skipped.
+        if not self.dynamic_loss_scale:
+            logger.warning(f"[deepspeed] OVERFLOW! Rank {dist.get_rank()} Skipping step. "
+                           "Gradients are non-finite, weights were not updated.")
```
[the remark this answers is not in the document]
Comment (sfc-gh-abkulkarni) [reply deepspeed/runtime/zero/stage3.py]: I am not extending the ZeRO-3 check to bf16 here: that would start skipping steps that currently proceed, which is a behavior change rather than a logging one and deserves its own issue and device test. Glad to file it if useful.

File: deepspeed/runtime/zero/stage_1_and_2.py [hunk 3]
```diff
@@ -2305,6 +2305,11 @@ def step(self, closure=None):
         prev_scale = self.loss_scale
         self._update_scale(self.overflow)
         if self.overflow:
+            # DynamicLossScaler reports the skip itself; a static scale (always the case for bf16)
+            # does not, leaving the step silently skipped.
+            if not self.dynamic_loss_scale:
+                logger.warning(f"[deepspeed] OVERFLOW! Rank {dist.get_rank()} Skipping step. "
+                               "Gradients are non-finite, weights were not updated.")
```
[the remark this answers is not in the document]
Comment (sfc-gh-abkulkarni) [reply deepspeed/runtime/zero/stage_1_and_2.py]: Fixed in 546dbb0 — guarded on `dist.get_rank() == 0`, matching `DynamicLossScaler`, and dropped the now-redundant rank from the message.

File: deepspeed/runtime/zero/stage3.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (sfc-gh-abkulkarni) [reply deepspeed/runtime/zero/stage3.py]: ZeRO-3 (and zenflow) call `check_overflow()` only for fp16, so bf16 never sets `self.overflow` and the step is not skipped there at all. That is a behavior change rather than a reporting one, so I have left it out and said so in the description.

File: deepspeed/runtime/zero/stage_1_and_2.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (sfc-gh-abkulkarni) [reply deepspeed/runtime/zero/stage_1_and_2.py]: Now handled by `log_dist(..., ranks=[0])` in 290275e, as suggested. That also turned out to be safer than a bare `dist.get_rank()`: `get_dist_msg` falls back gracefully when comm is not initialized, whereas `get_rank()` asserts.

**Review by pengdurice — APPROVED** [review]
